File: frontend/src/api/weekCategory.ts
import { apiFetch } from "./client"

export type WeekCategory =
  | "build"
  | "maintain"
  | "taper"
  | "race"
  | "restitution"
  | "recovering"

export interface WeekCategoryEntry {
  athlete_id: number
  week_start: string
  category: WeekCategory
  created_at: string
  updated_at: string
}

export function fetchWeekCategories(
  athleteId: number,
  start: string,
  end: string
): Promise<WeekCategoryEntry[]> {
  return apiFetch<WeekCategoryEntry[]>(
    `/api/v1/athlete/${athleteId}/week-category?start=${start}&end=${end}`
  )
}

export function setWeekCategory(
  athleteId: number,
  weekStart: string,
  category: WeekCategory
): Promise<WeekCategoryEntry> {
  return apiFetch<WeekCategoryEntry>(`/api/v1/athlete/${athleteId}/week-category`, {
    method: "POST",
    body: JSON.stringify({ week_start: weekStart, category }),
  })
}

File: sidekick/src/models/week_category.py
from datetime import datetime, timezone
from enum import Enum

from pydantic import AwareDatetime, BaseModel, Field, field_validator

from utils.datetime_utils import ensure_utc


class WeekCategory(str, Enum):
    BUILD = "build"
    MAINTAIN = "maintain"
    TAPER = "taper"
    RACE = "race"
    RESTITUTION = "restitution"
    RECOVERING = "recovering"


class WeekCategoryEntry(BaseModel):
    athlete_id: int
    week_start: str  # YYYY-MM-DD, Monday of the week
    category: WeekCategory
    created_at: AwareDatetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: AwareDatetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    @field_validator("created_at", "updated_at", mode="before")
    @classmethod
    def _utc(cls, v):
        return ensure_utc(v)


class WeekCategoryEntryRequest(BaseModel):
    week_start: str  # YYYY-MM-DD, Monday of the week
    category: WeekCategory
